Add logging when fail to kill container or pod

brianpursley opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
While investigating #93925, I noticed there are a couple places where killing a container or pod might fail without a message being logged. This PR adds logging when a container or pod fails to be killed.

**Which issue(s) this PR fixes**:


**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

brianpursley commented:
/retest

brianpursley commented:
/assign @yujuhong

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -1524,6 +1524,8 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 			if err := kl.killPod(pod, nil, podStatus, nil); err == nil {
 				podKilled = true
+			} else {
+				klog.Errorf("killPod for pod %q failed: %v", format.Pod(pod), err)
```
I wonder if in this case it should return as well. Alternative is that pod is terminating, runtime failed to kill the pod, and  will not be restarted.

Also, interesting thing here is that condition was added https://github.com/kubernetes/kubernetes/pull/39606 seemingly for the case when `podStatus` is `nil`. So it will not even attempt to call into the runtime

SergeyKanzhelev commented:
/unassign @yujuhong
/assign @dashpole

brianpursley commented:
I went ahead and updated this PR with the changes I think are correct. 

If they are not, I can update this PR, but I think it seems right to not kill the sandbox until the containers are killed.  I updated the description accordingly.

brianpursley commented:
@SergeyKanzhelev I certainly think there is some opportunity to clean up the pod/container kill code.

I'm not confident enough that changing the behavior will be OK at this point, which is in part why I would like to add some logging... to see if there are failures to kill pods/containers or not.

I can follow-on with another PR to change the behavior to return, but I would like to get some extra eyes on that change.

SergeyKanzhelev commented on pkg/kubelet/kubelet.go:
```diff
@@ -1524,6 +1524,8 @@ func (kl *Kubelet) syncPod(o syncPodOptions) error {
 			if err := kl.killPod(pod, nil, podStatus, nil); err == nil {
 				podKilled = true
+			} else {
+				klog.Errorf("killPod for pod %q failed: %v", format.Pod(pod), err)
```
as you picking the route of Can you also log the `podStatus`. I'm still concerned that this `if` was introduced in the PR #39606 that is checking for `podStatus` to be `nil`.

brianpursley replied:
Sure, I updated the log message to include podStatus

SergeyKanzhelev commented:
> @SergeyKanzhelev I certainly think there is some opportunity to clean up the pod/container kill code.
> 
> I'm not confident enough that changing the behavior will be OK at this point, which is in part why I would like to add some logging... to see if there are failures to kill pods/containers or not.
> 
> I can follow-on with another PR to change the behavior to return, but I would like to get some extra eyes on that change.

Do you have an environment where you can try to repro the bug that initiated this PR? If so, this gradual approach totally makes sense.

brianpursley commented:
> > @SergeyKanzhelev I certainly think there is some opportunity to clean up the pod/container kill code.
> > I'm not confident enough that changing the behavior will be OK at this point, which is in part why I would like to add some logging... to see if there are failures to kill pods/containers or not.
> > I can follow-on with another PR to change the behavior to return, but I would like to get some extra eyes on that change.
> 
> Do you have an environment where you can try to repro the bug that initiated this PR? If so, this gradual approach totally makes sense.

Not sure if I can reproduce it, but I can try in a local cluster.

brianpursley committed:
Add logging if container or pod fails to be killed

SergeyKanzhelev commented:
/retest
/lgtm
/assign @dashpole

dchen1107 commented:
/lgtm
/approve
